File: spring_java/src/main/java/dev/pollito/spring_java/config/security/handler/CustomAuthenticationEntryPoint.java
package dev.pollito.spring_java.config.security.handler;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import org.jspecify.annotations.NonNull;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.stereotype.Component;

@Component
public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {

  @Override
  public void commence(
      @NonNull HttpServletRequest request,
      @NonNull HttpServletResponse response,
      @NonNull AuthenticationException authException)
      throws IOException {
    AuthenticationErrorResponseWriter.write(
        request, response, HttpStatus.UNAUTHORIZED, authException.getMessage());
  }
}

File: spring_kotlin/src/main/kotlin/dev/pollito/spring_kotlin/config/security/handler/CustomAuthenticationEntryPoint.kt
package dev.pollito.spring_kotlin.config.security.handler

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus.UNAUTHORIZED
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.stereotype.Component

@Component
class CustomAuthenticationEntryPoint : AuthenticationEntryPoint {

  override fun commence(
      request: HttpServletRequest,
      response: HttpServletResponse,
      authException: AuthenticationException,
  ) {
    AuthenticationErrorResponseWriter.write(
        request,
        response,
        UNAUTHORIZED,
        authException.message ?: "Unauthorized",
    )
  }
}
